shell/platform/linux/fl_method_response.cc

```diff
@@ -0,0 +1,184 @@
+  FlValue* details;
+};
+
+struct _FlMethodNotImplementedResponse {
+  FlMethodResponse parent_instance;
+};
+
+G_DEFINE_TYPE(FlMethodResponse, fl_method_response, G_TYPE_OBJECT)
+G_DEFINE_TYPE(FlMethodSuccessResponse,
+              fl_method_success_response,
+              fl_method_response_get_type())
+G_DEFINE_TYPE(FlMethodErrorResponse,
+              fl_method_error_response,
+              fl_method_response_get_type())
+G_DEFINE_TYPE(FlMethodNotImplementedResponse,
+              fl_method_not_implemented_response,
+              fl_method_response_get_type())
+
+static void fl_method_response_class_init(FlMethodResponseClass* klass) {}
+
+static void fl_method_response_init(FlMethodResponse* self) {}
+
+static void fl_method_success_response_dispose(GObject* object) {
+  FlMethodSuccessResponse* self = FL_METHOD_SUCCESS_RESPONSE(object);
+
+  g_clear_pointer(&self->result, fl_value_unref);
+
+  G_OBJECT_CLASS(fl_method_success_response_parent_class)->dispose(object);
+}
+
+static void fl_method_success_response_class_init(
+    FlMethodSuccessResponseClass* klass) {
+  G_OBJECT_CLASS(klass)->dispose = fl_method_success_response_dispose;
+}
+
+static void fl_method_success_response_init(FlMethodSuccessResponse* self) {}
+
+static void fl_method_error_response_dispose(GObject* object) {
+  FlMethodErrorResponse* self = FL_METHOD_ERROR_RESPONSE(object);
+
+  g_clear_pointer(&self->code, g_free);
+  g_clear_pointer(&self->message, g_free);
+  g_clear_pointer(&self->details, fl_value_unref);
+
+  G_OBJECT_CLASS(fl_method_error_response_parent_class)->dispose(object);
+}
+
+static void fl_method_error_response_class_init(
+    FlMethodErrorResponseClass* klass) {
+  G_OBJECT_CLASS(klass)->dispose = fl_method_error_response_dispose;
+}
+
+static void fl_method_error_response_init(FlMethodErrorResponse* self) {}
+
+static void fl_method_not_implemented_response_class_init(
+    FlMethodNotImplementedResponseClass* klass) {}
+
+static void fl_method_not_implemented_response_init(
+    FlMethodNotImplementedResponse* self) {}
+
+G_MODULE_EXPORT FlValue* fl_method_response_get_result(FlMethodResponse* self,
+                                                       GError** error) {
+  if (FL_IS_METHOD_SUCCESS_RESPONSE(self))
+    return fl_method_success_response_get_result(
+        FL_METHOD_SUCCESS_RESPONSE(self));
+
+  if (FL_IS_METHOD_ERROR_RESPONSE(self)) {
+    const gchar* code =
+        fl_method_error_response_get_code(FL_METHOD_ERROR_RESPONSE(self));
+    const gchar* message =
+        fl_method_error_response_get_message(FL_METHOD_ERROR_RESPONSE(self));
+    FlValue* details =
+        fl_method_error_response_get_details(FL_METHOD_ERROR_RESPONSE(self));
+    g_autofree gchar* details_text = nullptr;
+    if (details != nullptr) {
+      FlJsonMessageCodec* codec = fl_json_message_codec_new();
+      g_autoptr(GError) json_error = nullptr;
+      details_text = fl_json_message_codec_encode(codec, details, &json_error);
```

Comment:
It feels like abusing the API to deliberately call JSON message codec—which is only supposed to be used with valid JSON structures—with arbitrary content and then potentially fail. It also means that a basically-random-to-users errors won't have details logged.

I would much prefer that FlValue implement a `to_string()` type method, and call that. It should be pretty simple to write, and will work with everything rather than just the things that happen to be valid json.

Reply:
I'll do this in a separate PR, dropping the JSON code for now.